File: Python/15.py
# 15. 3Sum

# Given an array nums of n integers,
#  are there elements a, b, c in nums such that a + b + c = 0?
#  Find all unique triplets in the array which gives the sum of zero.

# Note:

# The solution set must not contain duplicate triplets.

# Example:

# Given array nums = [-1, 0, 1, 2, -1, -4],

# A solution set is:
# [
#   [-1, 0, 1],
#   [-1, -1, 2]
# ]
from typing import List


class Solution(object):
    def threeSum(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        result = []
        length = len(nums)
        nums.sort()  # if nums[0] > 0: # return result
        for i in range(length - 2):
            #  same as formar
            if i > 0 and nums[i] == nums[i - 1]:
                continue
            left = i + 1
            right = length - 1
            while left < right:
                _sum = nums[left] + nums[i] + nums[right]
                if _sum == 0:
                    result.append([nums[left], nums[i], nums[right]])
                    left += 1
                    right -= 1
                    # same as formar -1
                    while left < right and nums[left] == nums[left - 1]:
                        left += 1
                    # same as formar +1
                    while left < right and nums[right] == nums[right + 1]:
                        right -= 1
                elif _sum < 0:
                    left += 1
                else:
                    right -= 1
        return result


if __name__ == "__main__":
    from util import Test

    t = Test(Solution().threeSum)
    t.equal([[-1, -1, 2], [-1, 0, 1]], [-1, 0, 1, 2, -1, -4])

File: cpluscplus/15.cpp
// 15. 3Sum
//
// Given an array nums of n integers,
//  are there elements a, b, c in nums
//  such that a + b + c = 0 ?
//  Find all unique triplets in the array
//  which gives the sum of zero.

// Given array nums = [ -1, 0, 1, 2, -1, -4 ],
//             A solution set is : [
//                 [ -1, 0, 1 ],
//                 [ -1, -1, 2 ]
//             ]
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

class Solution
{
public:
    vector<vector<int>> threeSum(vector<int> &nums)
    // for 0 to length-2
    // if same as former: contine
    // use two pointer left and right
    // find sum = equals zero add result to results
    // contine to find (remember skip the same value for both side!)
    //  if less than zero increase the left
    //  if great than zero decrease the right
    {
        vector<vector<int>> results = {};
        if (nums.size() < 2)
        {
            return results;
        }
        sort(nums.begin(), nums.end());
        for (int i = 0; i < nums.size() - 2; i++)
        {
            if (i > 0 && nums[i] == nums[i - 1])
            {
                continue;
            }
            int left = i + 1;
            int right = nums.size() - 1;
            while (left < right)
            {
                int sum = nums[i] + nums[left] + nums[right];
                if (sum == 0)
                {
                    results.push_back({nums[i], nums[left], nums[right]});
                    left++;
                    right--;
                    while (left < right && nums[left] == nums[left - 1])
                    {
                        left++;
                    }
                    while (left < right && nums[right] == nums[right + 1])
                    {
                        right--;
                    }
                }
                else if (sum < 0)
                {
                    left++;
                }
                else
                // sum greate than zero, decrease right
                {
                    right--;
                }
            }
        }
        return results;
    }
};

void show_2d_array(vector<vector<int>> ans)
{
    std::cout << "[ ";
    for (vector<int> l : ans)
    {
        std::cout << "[ ";
        for (int n : l)
        {
            std::cout << n << ", ";
        }
        std::cout << " ],";
    }
    std::cout << " ]" << std::endl;
}

int main()
{
    Solution s = Solution();
    vector<int> nums = {-1, 0, 1, 2, -1, -4};
    vector<vector<int>> ans = s.threeSum(nums);
    show_2d_array(ans);
    nums = {};
    ans = s.threeSum(nums);
    show_2d_array(ans);
    nums = {0};
    ans = s.threeSum(nums);
    show_2d_array(ans);
    nums = {-1, 0, 1, 2, -1, -4};
    ans = s.threeSum(nums);
    show_2d_array(ans);

    return 0;
}